addons/nuxeo-template-rendering/nuxeo-template-rendering-core/src/main/java/org/nuxeo/template/processors/AbstractBindingResolver.java

```diff
@@ -70,106 +78,179 @@ protected DocumentObjectWrapper getWrapper() {
-                            }
-                        }
+                Object value = extractValueFromParam(templateBasedDocument, param);
+                context.put(param.getName(), value);
+            } catch (ValueNotFound e) {
+                log.warn("Unable to handle binding for param " + param.getName(), e);
+            } catch (NoValueToAddInContext e) {
+                log.debug("Skip param to add " + param.getName(), e);
+            }
+        }
+    }
+
+    protected Object extractValueFromParam(TemplateBasedDocument templateBasedDocument, TemplateInput param) throws ValueNotFound, NoValueToAddInContext {
+        DocumentModel doc = templateBasedDocument.getAdaptedDoc();
+        String propKey = param.getSource();
+
+        switch (param.getType()) {
+            case BooleanValue:
+                return param.getBooleanValue();
+            case DateValue:
+                return param.getDateValue();
+            case StringValue:
+                return param.getStringValue();
+            case MapValue:
+                Map<String, Object> resultMap = new HashMap<>();
+                Map<String, TemplateInput> map = param.getMapValue();
+                for (String paramName : map.keySet()) {
+                    try {
+                        resultMap.put(paramName, extractValueFromParam(templateBasedDocument, map.get(paramName)));
+                    } catch (NoValueToAddInContext | ValueNotFound e) {
+                        log.debug("Skip param to add " + paramName + " in " + param.getName(), e);
                     }
-                    Property property = null;
+                }
+                return resultMap;
+            case ListValue:
+                List<Object> resultList = new ArrayList<>();
+                for (TemplateInput childParam : param.getListValue()) {
                     try {
-                        property = templateBasedDocument.getAdaptedDoc().getProperty(param.getSource());
-                    } catch (PropertyException e) {
-                        log.warn("Unable to ready property " + param.getSource(), e);
+                        resultList.add(extractValueFromParam(templateBasedDocument, childParam));
+                    } catch (NoValueToAddInContext | ValueNotFound e) {
+                        log.debug("Skip param to add " + childParam.getName() + " in " + param.getName(), e);
                     }
-
-                    Serializable value = null;
-                    if (property != null) {
-                        value = property.getValue();
+                }
+                return resultList;
+            case Content:
+                ContentInputType contentInput = ContentInputType.getByValue(param.getSource());
+                if (BlobContent.equals(contentInput))
+                    return extractBlobContent(doc, param);
+                else if (HtmlPreview.equals(contentInput))
```

Comment:
Please add braces everywhere in `if` and `else` statements (also in several other places in this class).